packages/client/test/rpc/engine/preimages.spec.ts

```diff
@@ -0,0 +1,190 @@
+import { getRpcClient, setupChain } from '../helpers.js'
+
+import type { Common } from '@ethereumjs/common'
+import type { PrefixedHexString } from '@ethereumjs/util'
+import type { HttpClient } from 'jayson/promise'
+
+const genesisStateRoot = '0x78026f1e4f2ff57c340634f844f47cb241beef4c965be86a483c855793e4b07d'
+const genesisBlockHash = '0x76a519ccb8a2b12d733ad7d88e2d5f4a11d6dc6ca320edccd3b8a3e9081ca1b3'
+
+const originalValidate = (BlockHeader as any).prototype._consensusFormatValidation
+BlockHeader.prototype['_consensusFormatValidation'] = () => {} //stub
+
+async function genBlockWithdrawals(blockNumber: number) {
+  // if block 1, bundle 0 withdrawals
+  const withdrawals =
+    blockNumber === 1
+      ? []
+      : Array.from({ length: 8 }, (_v, i) => {
+          const withdrawalIndex = blockNumber * 16 + i
+
+          // just return a withdrawal based on withdrawalIndex
+          return {
+            index: intToHex(withdrawalIndex),
+            validatorIndex: intToHex(withdrawalIndex),
+            address: bytesToHex(setLengthRight(intToBytes(withdrawalIndex), 20)),
+            amount: intToHex(withdrawalIndex),
+          }
+        })
+  const withdrawalsRoot = bytesToHex(
+    await Block.genWithdrawalsTrieRoot(withdrawals.map(Withdrawal.fromWithdrawalData))
+  )
+
+  return { withdrawals, withdrawalsRoot }
+}
+
+async function runBlock(
+  { common, rpc }: { common: Common; rpc: HttpClient },
+  runData: {
+    parentHash: PrefixedHexString
+    transactions: PrefixedHexString[]
+    blockNumber: PrefixedHexString
+    stateRoot: PrefixedHexString
+    receiptTrie: PrefixedHexString
+    gasUsed: PrefixedHexString
+    coinbase: PrefixedHexString
+  }
+) {
+  const { transactions, parentHash, blockNumber, stateRoot, receiptTrie, gasUsed, coinbase } =
+    runData
+  const txs = []
+  for (const [index, serializedTx] of transactions.entries()) {
+    try {
+      const tx = TransactionFactory.fromSerializedData(hexToBytes(serializedTx), {
+        common,
+      })
+      txs.push(tx)
+    } catch (error) {
+      const validationError = `Invalid tx at index ${index}: ${error}`
+      throw validationError
+    }
+  }
+  const transactionsTrie = bytesToHex(await Block.genTransactionsTrieRoot(txs))
+
+  const { withdrawals, withdrawalsRoot } = await genBlockWithdrawals(Number(blockNumber))
+
+  const headerData = {
+    parentHash,
+    number: blockNumber,
+    withdrawalsRoot,
+    transactionsTrie,
+    stateRoot,
+    receiptTrie,
+    gasUsed,
+    coinbase,
+  }
+  const blockData = { header: headerData, transactions: txs, withdrawals }
+  const executeBlock = Block.fromBlockData(blockData, { common })
+  const executePayload = blockToExecutionPayload(executeBlock, BigInt(0)).executionPayload
+  const res = await rpc.request('engine_newPayloadV2', [executePayload])
+  assert.equal(res.result.status, 'VALID', 'valid status should be received')
+  return executePayload
+}
+
```

Comment:
Just putting this a bit random in the test file, comment is for the whole test setup though:

Are these tests in any form preimage related? 🤔 If not then I would suggest to move to `verkle.spec.ts` (but I might be missing something).

The whole test setup code is pretty extensive, is there a way to consolidate this in an existing `util.ts` file or - otherway around - use existing setup functionality for getting the setup, so that we do not create that much code bloat here?

Reply:
I agree here, I do not see how this tests the preimages.

Reply:
ahh yes, the comparison with generated and expected pre0images is missing , can add that as well cc @gabrocheleau

Reply:
I believe this has been addressed with the latest changes.